### PR TITLE
[WIP] [Console] Refactored console formatting

### PR DESCRIPTION
**Pull request by fchris82**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #30519, #30920
| License       | MIT
| Doc PR        | TODO

Antecedent:
- https://github.com/symfony/symfony/pull/30519
- https://github.com/symfony/symfony/pull/30590

I made a full refactoring. The original output formatting isn't elegant and hard to extend. I made a lexer with tokens and visitors. This solution is based on Design Patterns: Elements of Reusable Object-Oriented Software book, from Erich Gamma, Richard Helm, Ralph Johnson, John M Vlissides (Chapter 2)

TODO:
- [ ] Extend Upgrade.md
- [ ] Create Doc PR
- [ ] Discussion

### CODE REVIEW

File: src/Symfony/Component/Console/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,13 @@ CHANGELOG
 
  * added support for hyperlinks
  * added `ProgressBar::iterate()` method that simplify updating the progress bar when iterating
+ * refactored console output formatter: add lexer and tokens. Important changes:
+    * added `<wrap>` and `<nowrap>` formatter tags to console
+    * [BC BREAK] earlier the unhandled/unknown tags will be shown, but now every tag will disappear after formatting. If you want to show a "tag", you have to escape it with `\` sign: `\<visibleTag>...\</visibleTag>`. You can use the `\Symfony\Component\Console\Formatter\OutputFormatter::escape()` function as well.
```
Comment (stof) [inline src/Symfony/Component/Console/CHANGELOG.md]: BC breaks are a no-go in a minor version. This must trigger a deprecation instead to provide a migration path.

**Comment by fchris82** [comment]
@ro0NL Hello again :) I'm waiting for your observations. Yes, it became huge. :-/

File: src/Symfony/Component/Console/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,13 @@ CHANGELOG
...
+    * [BC BREAK] `\Symfony\Component\Console\Formatter\OutputFormatterStyleInterface` was changed: `apply()` function was removed, and there were set 2 new methods: `start()`, `close()`
```
Comment (stof) [inline src/Symfony/Component/Console/CHANGELOG.md]: no-go. This must use deprecations, and adding new methods in an interface is also a no-go (use a separate interface maybe ?)

File: src/Symfony/Component/Console/CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,13 @@ CHANGELOG
...
+    * [BC BREAK] `\Symfony\Component\Console\Formatter\OutputFormatter` is no longer instance of `WrappableOutputFormatterInterface`
```
Comment (stof) [inline src/Symfony/Component/Console/CHANGELOG.md]: that BC break is also a no-go.

File: src/Symfony/Component/Console/Formatter/Lexer.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Tokenize a custom text to handle styling and wrapping it.
+ *
+ * @author Krisztián Ferenczi <ferenczi.krisztian@gmail.com>
+ */
+class Lexer implements LexerInterface
+{
+    /** @var FullTextToken */
+    protected $fullTextToken;
```
Comment (stof) [inline src/Symfony/Component/Console/Formatter/Lexer.php]: everything must use private visibility, not protected one, unless we *explicitly* decide to make inheritance the supported extension point (hint: such decision is always the last choice).

File: src/Symfony/Component/Console/Formatter/LexerInterface.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * @author Krisztián Ferenczi <ferenczi.krisztian@gmail.com>
+ */
+interface LexerInterface
+{
+    /**
+     * Tokenize a text.
+     *
+     * @param string $text
+     *
+     * @return \IteratorAggregate
```
Comment (stof) [inline src/Symfony/Component/Console/Formatter/LexerInterface.php]: should return `\Traversable` or `iterable`. There is no reason to enforce that `\IteratorAggregate` is used to implementator the iterator.

File: src/Symfony/Component/Console/Formatter/OutputFormatterStyle.php [hunk 6]
```diff
@@ -180,38 +174,55 @@ public function setOptions(array $options)
      * @param string $text The text to style
      *
      * @return string
+     *
+     * @deprecated
```
Comment (stof) [inline src/Symfony/Component/Console/Formatter/OutputFormatterStyle.php]: deprecation warnings must be triggered

File: src/Symfony/Component/Console/Formatter/OutputFormatterStyle.php [hunk 7]
```diff
@@ -120,11 +119,6 @@ public function setBackground($color = null)
         $this->background = static::$availableBackgroundColors[$color];
     }
 
-    public function setHref(string $url): void
```
Comment (stof) [inline src/Symfony/Component/Console/Formatter/OutputFormatterStyle.php]: removing that is a no-go.

File: src/Symfony/Component/Console/Formatter/OutputFormatterStyleStack.php [hunk 8]
```diff
@@ -42,32 +43,52 @@ public function reset()
 
     /**
      * Pushes a style in the stack.
+     *
+     * @param int                           $depth
+     * @param OutputFormatterStyleInterface $style
      */
-    public function push(OutputFormatterStyleInterface $style)
+    public function push(int $depth, OutputFormatterStyleInterface $style)
```
Comment (stof) [inline src/Symfony/Component/Console/Formatter/OutputFormatterStyleStack.php]: this is a BC break

File: src/Symfony/Component/Console/Formatter/OutputFormatterStyleStack.php [hunk 9]
```diff
@@ -42,32 +43,52 @@ public function reset()
...
     {
-        $this->styles[] = $style;
+        $this->styles[$depth] = $style;
```
Comment (stof) [inline src/Symfony/Component/Console/Formatter/OutputFormatterStyleStack.php]: this is not pushing anything anymore. It replaces an element of the stack

File: src/Symfony/Component/Console/Formatter/Tokens/EosToken.php [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Console\Formatter\Tokens;
```
Comment (stof) [inline src/Symfony/Component/Console/Formatter/Tokens/EosToken.php]: our namespaces are singular

File: src/Symfony/Component/Console/Tests/Formatter/OutputFormatterTest.php [hunk 11]
```diff
@@ -15,6 +15,10 @@
 use Symfony\Component\Console\Formatter\OutputFormatter;
 use Symfony\Component\Console\Formatter\OutputFormatterStyle;
 
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ * @author Krisztián Ferenczi <ferenczi.krisztian@gmail.com>
```
Comment (stof) [inline src/Symfony/Component/Console/Tests/Formatter/OutputFormatterTest.php]: We don't put `@author` phpdoc on test files.

File: src/Symfony/Component/Console/Tests/Formatter/OutputFormatterTest.php [hunk 12]
```diff
@@ -124,10 +128,9 @@ public function testDeepNestedStyles()
 
     public function testNewStyle()
     {
-        $formatter = new OutputFormatter(true);
-
         $style = new OutputFormatterStyle('blue', 'white');
-        $formatter->setStyle('test', $style);
```
Comment (stof) [inline src/Symfony/Component/Console/Tests/Formatter/OutputFormatterTest.php]: this test should be kept unchanged. If you cannot, it means there is a BC break. And that's not something we want

File: src/Symfony/Component/Console/Tests/Formatter/OutputFormatterTest.php [hunk 13]
```diff
@@ -318,26 +323,13 @@ public function testContentWithLineBreaks()
         ));
     }
 
-    public function testFormatAndWrap()
```
Comment (stof) [inline src/Symfony/Component/Console/Tests/Formatter/OutputFormatterTest.php]: this test must be kept.

**Comment by stof** [comment]
Can we also get a performance comparison of the old and new system ?

**Comment by fchris82** [comment]
> Can we also get a performance comparison of the old and new system ?

I will make one. I think the new is worse, but I thought this is the console, we don't want to format large texts/outputs, and the time isn't so important on the console. I inserted a limitation here: https://github.com/symfony/symfony/pull/31131/files#diff-a42222f449d2d525a1100ed2b995e429R43 on the bases of this: http://piotrpasich.com/spl-iterators-against-the-performance/ (first comment: http://piotrpasich.com/spl-iterators-against-the-performance/#comment-1818526630 )

**Comment by fchris82** [comment]
@stof I wrote a benchmark: https://gist.github.com/fchris82/e09e4d9723e4b0e73a768ab53860d321 and I ran it:

```
1000x original
--------------
Short           0.03s
Short with tags 0.29s
Simple          0.03s
With tags       1.56s

100x new
--------
Short           1.02s
Short with tags 1.24s
Simple          44.41s
With tags       47.42s
```

Yes, there are big differences. Maybe I should keep the original Formatter and create a new one for wrapping and other actions?

**Comment by fchris82** [comment]
@stof Hello, I made a twice faster solution with plain array token container: https://github.com/fchris82/symfony/commit/d4fc1d2b4caf444f3030342022dd577f047f4e92 
```php
$tokens = [
    [Lexer::TYPE_WORD, 'Lorem'],
    [Lexer::TYPE_SEPARATOR, ' '],
    [Lexer::TYPE_WORD, 'ipsum'],
    // ...
    new FullTokenTag('<comment>'),
    // ...
];
```
This solution is faster, but some things are lost in this case, eg: type hinting (there are strings and Token objects mixed in "one array") What do you prefer?

My opinion: I kept the original `OutputFormatter` also, so if somebody want to format anything fast (eg: large text and wrapping insn't important), original can be used for them. In most of case we use short texts on console, "twice time" doesn't matter.

**Comment by stof** [comment]
why does your benchmark result have `1000x original` and `100x new` ? We need *comparable* data. If I understand things properly, the old code can format 1000 messages in 30ms, meaning 30µs per message, while the new code formats 100 messages in 1s, meaning 10000µs per message. That's not twice slower. That's 300 times slower.

**Comment by fchris82** [comment]
@stof You missunderstood my last comment. There are more problems with the word wrapping in SymfonyStyle now:
1. tag wrapping
2. UTF-8 chars
3. wrapping URL-s

There are 3 different "solution" now:
- A fast fix: https://github.com/symfony/symfony/pull/30519 and https://github.com/symfony/symfony/pull/30590 ("speed": ~1-2x current)
- "this" ("speed": 300x current)
- like this, just a little bit optimazed:  https://github.com/fchris82/symfony/commit/d4fc1d2b4caf444f3030342022dd577f047f4e92 <-- an other branch! ("speed": 150x current)

Yes, this solution is much slower, but **works**, wrap the text well, easy to extend, and my question is: does this speed matter on console? Is there any different than the full text is shown `0.03 s` instead of `0.0001 s` on the screen? Important: we are talking about console, we use here shorter texts. (What I used in the performance test file, it is a long lorem ipsum text, we almost never print like this) If you think, that is too slow, I will finish the first version. If you think, this can be acceptable, I will finish this version.

I would like solve the problem, I don't insist on this version, only I think this is the "best" version, with an reasonable speed "problem".

**Comment by fchris82** [comment]
@stof Please give me a direction, what do you prefer:
1. fast with double "parsing"
2. slow but easy to extend

**Pull request closed without merging**